### PR TITLE
Refactor cleanup of file references

### PR DESCRIPTION
**Pull request by hmusum**

### CODE REVIEW

File: config-proxy/src/test/java/com/yahoo/vespa/config/proxy/filedistribution/CachedFilesMaintainerTest.java [hunk 1]
```diff
@@ -80,4 +85,16 @@ private static File newFolder(File root, String... subDirs) throws IOException {
...
     }
 
+    private void createFiles() {
+        IntStream.of(0,1,2,3).forEach(i -> {
+            try {
+                writeFileAndSetLastAccessedTime(cachedFileReferences, "fileReference" + i);
+                writeFileAndSetLastAccessedTime(cachedDownloads, "download" + i);
+                clock.advance(Duration.ofMinutes(1));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Comment (hakonhall) [inline config-proxy/src/test/java/com/yahoo/vespa/config/proxy/filedistribution/CachedFilesMaintainerTest.java:95]: Throw UncheckedIOException?

File: filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Returns all files in the given directory, non-recursive.
+     */
+    public static Set<String> getFileReferencesOnDisk(File directory) {
+        Set<String> fileReferencesOnDisk = new HashSet<>();
+        File[] filesOnDisk = directory.listFiles();
```
Comment (hakonhall) [inline filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java:103]: This has been known to cause OutOfMemoryError if the directory contains enough entries.  Consider reimplementing this, see UnixPath::listContentsOfDirectory.

**Review by hakonhall — APPROVED** [review]

File: config-proxy/src/test/java/com/yahoo/vespa/config/proxy/filedistribution/CachedFilesMaintainerTest.java (hunk 1 above)
Comment (hmusum) [reply config-proxy/src/test/java/com/yahoo/vespa/config/proxy/filedistribution/CachedFilesMaintainerTest.java:95]: Yes, good point, will do

File: filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java (hunk 2 above)
Comment (hmusum) [reply filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java:103]: Yeah, will do something similar

File: filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    public List<String> deleteUnusedFileReferences(File fileReferencesPath,
+                                                   Duration keepFileReferencesDuration,
+                                                   int numberToAlwaysKeep,
+                                                   Set<String> fileReferencesInUse) {
+        log.log(Level.FINE, () -> "Keep unused file references for " + keepFileReferencesDuration);
+        if (!fileReferencesPath.isDirectory()) throw new RuntimeException(fileReferencesPath + " is not a directory");
+
+        log.log(Level.FINE, () -> "File references in use : " + fileReferencesInUse);
+
+        Stream<String> candidates = sortedUnusedFileReferences(fileReferencesPath.toPath(), fileReferencesInUse, keepFileReferencesDuration);
```
Comment (hakonhall) [inline filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java:60]: Add a try-with-resources to ensure the directory file descriptor is finally closed.

Comment (hmusum) [reply filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java]: Thanks, fixed

**Review by hakonhall — APPROVED** [review]

**Pull request merged**